packages/client/src/components/SafeSettings/components/EditThreshold/index.js

```diff
@@ -1,63 +1,103 @@
-import { useContext } from "react";
+import { useState, useContext } from "react";
+import ReviewUpdateThreshold from "./ReviewUpdateThreshold.js";
+import EditThresholdForm from "./EditThresholdForm";
+import ReviewEditSafeOwners from "../Owners/ReviewEditSafeOwners";
+import { Web3Context } from "contexts/Web3";
 import { useHistory } from "react-router-dom";
+import { useModalContext } from "contexts/Modal";
 import { formatAddress } from "utils";
-import { useState } from "react";
-import { useModalContext } from "contexts";
-import { Web3Context } from "contexts/Web3";
-import ReviewSafeEdits from "./ReviewSafeEdits";
-import EditThresholdForm from "./EditThresholdForm";
-const EditThreshold = ({ treasury, newOwner }) => {
+
+const EditThreshold = ({ treasury, newOwner, ownerToBeRemoved }) => {
+  const { openModal, closeModal } = useModalContext()
   const history = useHistory();
-  const { closeModal } = useModalContext();
-  const [currStep, setCurrStep] = useState(0);
-  const { safeOwners, threshold, address } = treasury;
-  const { proposeAddSigner, updateThreshold, setTreasury } =
-    useContext(Web3Context);
-  const [newThreshold, setNewThreshold] = useState(Number(threshold));
+  const web3 = useContext(Web3Context);
+  const { setTreasury, proposeAddSigner, proposeRemoveSigner } = web3;
+  const { address, safeOwners, threshold } = treasury;
   const verifiedSafeOwners = safeOwners.filter((o) => o.verified);
-  const allSafeOwners = newOwner
-    ? [...verifiedSafeOwners, newOwner]
-    : verifiedSafeOwners;
+  let allSafeOwners;
+  if (newOwner) {
+    allSafeOwners = [...verifiedSafeOwners, newOwner];
+  } else if (ownerToBeRemoved) {
+    allSafeOwners = verifiedSafeOwners.filter(o => o.address !== ownerToBeRemoved.address);
+  } else {
+    allSafeOwners = verifiedSafeOwners
+  }
+  const [newThreshold, setNewThreshold] = useState(Math.min(Number(threshold), allSafeOwners.length));
+  
   const canContinueToReview = !!newOwner || newThreshold !== threshold;
   const onChangeThreshold = (isIncrease) => {
     setNewThreshold((prevState) =>
       isIncrease ? prevState + 1 : prevState - 1
     );
   };
 
-  const onSubmitClick = async () => {
+  const onConfirmAddOwner = async () => {
+    await proposeAddSigner(formatAddress(newOwner.address), newThreshold);
+    setTreasury(address, {
+      safeOwners: [...safeOwners, { ...newOwner, verified: false }],
+    })
+    closeModal();
+    history.push(`/safe/${address}`);
+  }
+
+  const onConfirmRemoveOwner = async () => {
+    await proposeRemoveSigner(formatAddress(ownerToBeRemoved.address), newThreshold);
+    setTreasury(address, treasury);
```

Comment:
looking at the proposeRemoveSigner function definition, looks like we are already refreshing the treasury after the transaction is sealed. We can remove this line now. Sorry for the confusion!